pkg/devicerepository/grpc.go

```diff
@@ -0,0 +1,247 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package devicerepository
+
+import (
+	"context"
+	"strconv"
+	"strings"
+
+	clusterauth "go.thethings.network/lorawan-stack/v3/pkg/auth/cluster"
+	"go.thethings.network/lorawan-stack/v3/pkg/auth/rights"
+	"go.thethings.network/lorawan-stack/v3/pkg/devicerepository/store"
+	"go.thethings.network/lorawan-stack/v3/pkg/errors"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// withDefaultModelFields appends default ttnpb.EndDeviceModel fields.
+func withDefaultModelFields(paths []string) []string {
+	return ttnpb.AddFields(paths, "brand_id", "model_id")
+}
+
+// withDefaultBrandFields appends default ttnpb.EndDeviceBrand paths.
+func withDefaultBrandFields(paths []string) []string {
+	return ttnpb.AddFields(paths, "brand_id")
+}
+
+func (dr *DeviceRepository) assetURL(brandID, path string) string {
+	if path == "" || dr.config.AssetsBaseURL == "" {
+		return path
+	}
+	return strings.TrimRight(dr.config.AssetsBaseURL, "/") + "/vendor/" + brandID + "/" + path
+}
+
+var (
+	errNoAuthentication = errors.DefineUnauthenticated("no_authentication", "no authentication")
+)
+
+// ListBrands implements the ttnpb.DeviceRepositoryServer interface.
+func (dr *DeviceRepository) ListBrands(ctx context.Context, req *ttnpb.ListEndDeviceBrandsRequest) (*ttnpb.ListEndDeviceBrandsResponse, error) {
+	if dr.config.RequireAuth {
+		if req.ApplicationIDs == nil {
+			return nil, errNoAuthentication.New()
+		}
+		if err := rights.RequireApplication(ctx, *req.ApplicationIDs, ttnpb.RIGHT_APPLICATION_DEVICES_READ); err != nil {
+			return nil, err
+		}
+	}
```

Comment:
What is the use case of _not_ requiring auth?

Reply:
Running a standalone Device Repository server, without the need to have an IS.

Reply:
There shall be an Identity Server in that case, with a mock application or anything.

Using TTS Device Repository component as a work-around for other use cases is fine, but with the strict precondition that it does not affect any code in The Things Stack, and certainly not security related.